util/HalideTraceDump.cpp

```diff
@@ -0,0 +1,408 @@
+                    max_coords[i] = p->coordinates()[lanes*i+lane];
+                }
+            }
+        }
+    }
+
+    void allocate() {
+        int size = 1;
+        for (int i = 0; i < dimensions; i++) {
+            extents[i] = max_coords[i] - min_coords[i] + 1;
+            size *= extents[i];
+        }
+        values = (Value *)malloc(size * sizeof(Value));
+        for (int i = 0; i < size; i++) {
+            values[i] = Value();
+        }
+        if (values == nullptr) {
+            fprintf(stderr, "Memory allocation failure. Aborting.\n");
+            exit(-1);
+        }
+    }
+
+    void add(Packet *p) {
+        int real_dims = p->dimensions/p->type.lanes;
+        int lanes = p->type.lanes;
+
+        if (values == nullptr) {
+            fprintf(stderr, "Packet storage not allocated. Aborting.\n");
+            exit(-1);
+        }
+        if (real_dims < 0) {
+            fprintf(stderr, "Negative dimensionality found. Aborting.\n");
+            exit(-1);
+        }
+
+        for (int lane = 0; lane < lanes; lane++) {
+            int offset = p->coordinates()[lane];
+            for (int i = 1; i < dimensions; i++) {
+                offset += (p->coordinates()[lanes*i+lane] * extents[i-1]);
+            }
+            values[offset] = Value(p, lane);
+        }
+    }
+
+    bool allocated() {
+        return (values != NULL);
+    }
+
+    template<typename T, int D>
+    Buffer<T, D> get_buffer() {
+        halide_buffer_t buf;
+        buf.device = 0;
+        buf.device_interface = nullptr;
+        buf.flags = 0;
+        buf.padding = nullptr;
+        buf.host = (uint8_t*)values;
+        halide_dimension_t dims[dimensions];
+        dims[0].min = 0;
+        dims[0].extent = extents[0];
+        dims[0].stride = sizeof(Value)/type.bytes();
+        for (int i = 1; i < dimensions; i++) {
+            dims[i].min = 0;
+            dims[i].extent = extents[i];
+            dims[i].stride = (dims[i-1].stride * dims[i-1].extent);
+        }
+        buf.dim = dims;
+        buf.dimensions = dimensions;
+        buf.type = type;
+
+        Buffer<T, D> buffer(buf);
+        return buffer;
+    }
+};
+
+void dump_func(string name, FuncInfo &func, BufferOutputOpts output_opts) {
+    // For now, support only 2D buffers.
+    if (func.dimensions != 2) {
+        printf("[INFO] Skipping func '%s', which is not 2D (has %d dimensions).\n", name.c_str(), func.dimensions);
+    }
+
+    string filename;
+    switch(output_opts.type) {
+    case BufferOutputOpts::PNG:
+        filename = name + ".png";
+        break;
+    case BufferOutputOpts::JPG:
+        filename = name + ".jpg";
+        break;
+    case BufferOutputOpts::PGM:
+        filename = name + ".pgm";
+        break;
+    default:
+        exit(1);
+    }
+
+    printf("[INFO] Dumping func '%s' to file: %s\n", name.c_str(), filename.c_str());
+
+    //Rely on save_image to do type-checking
+    auto buf = func.get_buffer<void, 2>();
```

Comment:
Actually, get_buffer doesn't need to be templatized at all. It can always return a Buffer<void>.